tmt/hardware.py

```diff
@@ -0,0 +1,1088 @@
+    """
+
+    def __init__(
+            self,
+            reducer: ReducerType = any,
+            constraints: Optional[List[ConstraintBase]] = None
+            ) -> None:
+        """
+        Construct a compound constraint, constraint imposed to more than one dimension.
+
+        :param reducer: a callable reducing a list of results from child constraints into the final
+            answer.
+        :param constraints: child contraints.
+        """
+
+        self.reducer = reducer
+        self.constraints = constraints or []
+
+    @property
+    def size(self) -> int:
+        return len(self.constraints)
+
+    def to_spec(self) -> Any:
+        return {
+            self.__class__.__name__.lower(): [
+                constraint.to_spec() for constraint in self.constraints
+                ]
+            }
+
+    def uses_constraint(self, parent: tmt.utils.Common, constraint_name: str) -> bool:
+        """
+        Inspect constraint whether it or its children use a constraint of a given name.
+
+        :param logger: logger to use for logging.
+        :param constraint_name: constraint name to look for.
+        :returns: ``True`` if the given constraint or its children use given constraint name.
+        """
+
+        # Using "any" on purpose: we cannot use the reducer belonging to this constraint,
+        # because that one may yield result based on validity of all child constraints.
+        # But we want to answer the question "is *any* of child constraints using the given
+        # constraint?", not "are all using it?".
+        return any(
+            constraint.uses_constraint(parent, constraint_name)
+            for constraint in self.constraints
+            )
+
+    def spans(
+            self,
+            members: Optional[List[ConstraintBase]] = None
+            ) -> Iterator[List[ConstraintBase]]:
+        """
+        Generate all distinct spans covered by this constraint.
+
+        Since the ``and`` reducer demands all child constraints must be satisfied, and some of
+        these constraints can also be compound constraints, we need to construct a cartesian
+        product of spans yielded by child constraints to include all possible combinations.
+
+        :param logger: logger to use for logging.
+        :param members: if specified, each span generated by this method is prepended with this
+            list.
+        :raises NotImplementedError: default implementation is left undefined for compound
+            constraints.
+        """
+
+        raise NotImplementedError
+
+
+@dataclasses.dataclass(repr=False)
+class Constraint(ConstraintBase):
+    """
+    A constraint imposing a particular limit to one of the guest properties.
+    """
+
+    # Name of the constraint. Used for logging purposes, usually matches the
+    # name of the system property.
+    name: str
+
+    # A binary operation to use for comparing the constraint value and the
+    # value specified by system or flavor.
```

Comment:
The `system or flavor` are a bit irrelevant here?

Reply:
Actually, I rephrased the sentence, and the meaning is still valid. The only user so far, in tmt repo, is the beaker plugin, which does no comparison but rewrites constraints into XML and leaves comparisons and filtering to Beaker. But for `virtual` or `podman`, the comparison will happen some day.

Reply:
Thanks for clarification.